src/data/buffer/parser/numeral.lean

```diff
@@ -0,0 +1,112 @@
+/-
+Copyright (c) 2020 Yakov Pechersky. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky
+-/
+import data.fintype.card
+
+/-!
+# Numeral parsers
+
+This file expands on the existing `nat : parser ℕ` to provide parsers into any type `α` that
+can be represented by a numeral, which relies on `α` having a 0, 1, and addition operation.
+There are also convenience parsers that ensure that the numeral parsed in is not larger than
+the cardinality of the type `α` , if it is known that `fintype α`. Additionally, there are
+convenience parsers that parse in starting from "1", which can be useful for positive ordinals;
+or parser from a given character or character range.
+
+## Main definitions
+
+* 'numeral` : The parser which uses `nat.cast` to map the result of `parser.nat` to the desired `α`
+* `numeral.of_fintype` :  The parser which `guard`s to make sure the parsed numeral is within the
+  cardinality of the target `fintype` type `α`.
+
+## Implementation details
+
+When the `numeral` or related parsers are invoked, the desired type is provided explicitly. In many
+cases, it can be inferred, so one can write, for example
+```lean
+def get_fin : string → fin 5 :=
+sum.elim (λ _, 0) id ∘ parser.run_string (parser.numeral.of_fintype _)
+```
+
+In the definitions of the parsers (except for `numeral`), there is an implicit `nat.cast` in
+the final `pure` statement.
+-/
+
+open parser parse_result
+
+namespace parser
+
+variables (α : Type) [has_zero α] [has_one α] [has_add α]
+
+/--
+Parse a string of digits as a numeral while casting it to target type `α`.
+-/
+def numeral : parser α :=
+nat.cast <$> nat
```

Comment:
This is a dangerous definition that should never be used in executable code because it has horrible performance.  `nat.cast n` is defined by (successor-)recursion on `n`.  Hence the runtime is *at least* linear in the *unary* representation of `n`.  Or exponential runtime in the usual representation.

I'm not sure where you want to go with this PR.  If you want to actually run the parsers on real strings, then this definition needs to be changed.
1. You could define `nat.bin_cast` using `nat.binary_rec` (plus a corresponding simp-lemma `n.bin_cast = n.cast`) and use it here.
2. Or you could define a Lean4-style `has_of_nat` type class (that contains an `of_nat` operation` together with a proof of `of_nat n = n.cast`).  I vaguely recall some discussion about such a class, but I can't find it now.

If you only want to verify the correctness of a parser (without ever running it), then the performance is of course irrelevant.  But then please put in a big warning sign in the documentation about the horrible performance.